lib/oxsdatatypes/src/parser.rs

```diff
@@ -125,412 +83,525 @@ impl From<DateTimeError> for XsdParseError {
     }
 }
 
-impl From<Err<Self>> for XsdParseError {
-    fn from(err: Err<Self>) -> Self {
-        match err {
-            Err::Incomplete(needed) => Self {
-                kind: XsdParseErrorKind::MissingData(needed),
-            },
-            Err::Error(e) | Err::Failure(e) => e,
-        }
-    }
+// [6]   duYearFrag ::= unsignedNoDecimalPtNumeral 'Y'
+// [7]   duMonthFrag ::= unsignedNoDecimalPtNumeral 'M'
+// [8]   duDayFrag ::= unsignedNoDecimalPtNumeral 'D'
+// [9]   duHourFrag ::= unsignedNoDecimalPtNumeral 'H'
+// [10]   duMinuteFrag ::= unsignedNoDecimalPtNumeral 'M'
+// [11]   duSecondFrag ::= (unsignedNoDecimalPtNumeral | unsignedDecimalPtNumeral) 'S'
+// [12]   duYearMonthFrag ::= (duYearFrag duMonthFrag?) | duMonthFrag
+// [13]   duTimeFrag ::= 'T' ((duHourFrag duMinuteFrag? duSecondFrag?) | (duMinuteFrag duSecondFrag?) | duSecondFrag)
+// [14]   duDayTimeFrag ::= (duDayFrag duTimeFrag?) | duTimeFrag
+// [15]   durationLexicalRep ::= '-'? 'P' ((duYearMonthFrag duDayTimeFrag?) | duDayTimeFrag)
+struct DurationParts {
+    year_month: Option<i64>,
+    day_time: Option<Decimal>,
 }
 
-type XsdResult<'a, T> = IResult<&'a str, T, XsdParseError>;
-
-const OVERFLOW_ERROR: XsdParseError = XsdParseError {
-    kind: XsdParseErrorKind::Overflow,
-};
-
-pub fn parse_value<'a, T>(
-    mut f: impl FnMut(&'a str) -> XsdResult<'a, T>,
-    input: &'a str,
-) -> Result<T, XsdParseError> {
-    let (left, result) = f(input)?;
-    if left.is_empty() {
-        Ok(result)
+fn duration_parts(input: &str) -> Result<(DurationParts, &str), XsdParseError> {
+    let (negative, input) = if let Some(left) = input.strip_prefix('-') {
+        (true, left)
     } else {
-        Err(XsdParseError {
-            kind: XsdParseErrorKind::TooMuchData { count: left.len() },
-        })
+        (false, input)
+    };
+    let mut input = expect_char(input, 'P', "Durations must start with 'P'")?;
+    let mut state = 0; // after_year = 1, after_month = 2, after_days = 3, after_T = 4, after_hours = 5, after_minutes = 6, after_seconds = 7
```

Comment:
I understand that an `enum`-based approach would make the less-than comparison below unergonomic, but maybe named constants for the values would be useful still?

Reply:
Indeed. It would be way more readable. Done. Thank you!